shader/flatten.wgsl

```diff
@@ -280,39 +280,95 @@ fn flatten_cubic(cubic: Cubic, local_to_device: Transform, offset: f32) {
     }
 }
 
+// Flattens the circular arc that subtends the angle begin-center-end. It is assumed that
```

Comment:
It's not obvious from this comment how the direction is chosen. It must be based on the value of `angle`, but it feels like 'pacman' style arcs will be difficult.

I worry that if n_lines is big, the arc would draw the full way around, but if n_lines is small, it would end up drawing the closing part, rather than the full way around

I'm not sure that this is coherent - my last maths education was many years ago.

Maybe this would only be an issue if n_lines is 1, which is impossible?

Reply:
The direction of the arc is always a counter-clockwise rotation starting from `begin`, towards `end`, centered at `center`, and will be subtended by `angle` (which is assumed to be positive). If `angle` terminates the arc before it smoothly reaches `end`, a line will be drawn from the end of the arc towards `end`. I will document this.

`n_lines` only depends on the radius of the arc and the chosen tolerance.  For `n_lines` to be $1$, `theta` must be close to $6.2831853$ ($2\pi$). `acos` can't return that [by definition](https://www.w3.org/TR/WGSL/#acos-builtin). _If_ you somehow got `n_lines == 1`, basically no circle would be drawn and you'd get a line connecting `begin` and `end`.

Realistically, `n_lines` could be as low as $2$ if the absolute value of `x` is very small, which can happen if `tol` and `radius` are close to each other (the code prevents `radius` from being smaller than `tol`, so that case isn't possible). You'd only get into that situation with a sub-pixel radius. At that scale, the coarseness of the approximation won't be easily noticeable. If you increased the tolerance to make this possible at larger radii, with `n_lines == 2` you'd get a shape resembling a wedge/triangle (I tested this locally, which works as you'd expect).

That said, this function is meant for drawing caps and joins where `angle`$\le\pi$ and the `begin`, `center`, and `end` points are chosen carefully (so that the angle and winding make sense). 

Though there is something that should get fixed: I think selecting `1u` when `theta <= EPS` is the wrong thing to do and this is a bug. That happens when the subdivision count tends to infinity. It's probably better to define a `MAX_LINES` and return that instead.

Reply:
To illustrate what I mean, here is what the code draws with a center directly below the join and an angle of $343.77\degree$:

- `n_lines == 1`:
<img width="285" alt="Screenshot 2023-11-17 at 12 10 01 PM" src="https://github.com/linebender/vello/assets/6933700/8ea9162f-0bec-4878-b433-11e6d4b140a5">

- `n_lines == 2`:
<img width="251" alt="Screenshot 2023-11-17 at 12 10 44 PM" src="https://github.com/linebender/vello/assets/6933700/68c7c672-493a-4b2a-a9a9-812f36f4b791">

- `n_lines == 3`:
<img width="251" alt="Screenshot 2023-11-17 at 12 11 29 PM" src="https://github.com/linebender/vello/assets/6933700/4c1aae0b-5457-4869-89bd-51954330465f">

- `n_lines == 6`:
<img width="328" alt="Screenshot 2023-11-17 at 12 12 18 PM" src="https://github.com/linebender/vello/assets/6933700/42bfdce1-80b1-48fb-b448-e165ad90dc34">

- `n_lines == 100`:
<img width="354" alt="Screenshot 2023-11-17 at 12 12 58 PM" src="https://github.com/linebender/vello/assets/6933700/9151ed1a-8ebc-4d35-9bac-cf3042b80d77">

Reply:
Thanks for the very detailed reply!

I hadn't clocked the additional invariants being maintained by the callers

Those pictures are very cool as well!

Reply:
I'm glad they were helpful 🙂. I updated the comment above `flatten_arc` in [latest commit](https://github.com/linebender/vello/pull/414/commits/6e6d02a69e4df6b297bedde16370053e05d6e207).